### PR TITLE
Preserve Data

### PR DESCRIPTION
**Pull request by Munsa1**
# Preserve Data

### In the following milestone we:
- Check the documentation for Ruby's [JSON library](https://ruby-doc.org/stdlib-3.0.0/libdoc/json/rdoc/JSON.html).
- On exit from our app saves all the data (books, people, and rentals) into JSON files.
- On startup of our app we load all data from files. 
- We handle the case when one or more of the files are missing.

### CODE REVIEW

File: datastorage.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      data << if person.instance_of?(Teacher)
+                { key: 'teacher', id: person.id, age: person.age, name: person.name,
+                  specialization: person.specialization }
+              else
+                { key: 'student', id: person.id, age: person.age, name: person.name,
+                  parent_permission: person.parent_permission, classroom: person.classroom }
+              end
+      save_data('person.json', data)
+    end
+  end
```
Comment (mikethreels) [inline datastorage.rb]: - [ ] if the user opens the application and adds data to all the columns, then closes it and tries to open the application again they receive below error.
Screenshot 2022-02-09 120956
when looking at the .json file of person after that the name and age are being switched around, it seems that is the problem.
Screenshot 2022-02-09 121436

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by dasileker — CHANGES_REQUESTED** [review]
## **Hi @Munsa1,** 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

- Hello nice job but Like my colleague said the app is not running after exit and re-opening, the data is preserved correctly but like he said after exit and re-opening the app is generating an error and that shouldn't happen.

Capture d’écran 2022-02-09 133837


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by dasileker — APPROVED** [review]
## **Hi @,** 

### **Greate Work 👌**

## **Congratulations! 🎉 Project Approved ✅**
Your project is complete!  There is nothing else to say other than... it's time to merge it 🚀 
 


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**